### ISSUE TITLE
Automatic token refresh fails when called from parallel calls. 

### ISSUE THREAD

**Issue by KrasimirGatevMentorMate** [auth0/Auth0.Android#540]
In our current project we have an AuthInterceptor class that automatically retrieves the user's access token using the Auth0 secure credentials manager and using it for our calls. The issue is when the access token has expired and Auth0 library tries to refresh it - if at that time 2 or more parallel calls are called by the app the refresh procedure will fail every time - as the Auth0 secure credentials manager library does not have a latch or suspend function so the refresh token becomes invalidated before the refresh can complete. The internal refresh process will be called twice leading to the refresh process - failing every time even started with a valid refresh token.

In short triggering a token refresh in the app while two or more parallel calls are in progress will lead to the token refresh and consequently the calls failing 100% of the time.

**Comment by poovamraj** [comment]
Hi @KrasimirGatevMentorMate, thanks for raising this. We are internally following up on a similar issue, I will check your specific case and if they are related you can expect a fix for this in one of our next releases. In the meantime, I will also check if there is any workaround that can fix this for you.

**Comment by KrasimirGatevMentorMate** [comment]
Thank you for your quick answer. We do not have all that much parallel calls in the app currently, so we can workaround this for now, but I will be monitoring for updates.

**Comment by poovamraj** [comment]
Hi @KrasimirGatevMentorMate, This is the same issue that has already been noted in other tickets like #540, #330, #359. We have noted in our README which mentions how our current version is not thread-safe. We are soon planning to start working on a version that will support thread-safe Credential Manager. I will keep you posted on this thread about its progress. Thanks for reporting this to us.

**Comment by poovamraj** [comment]
Hi @KrasimirGatevMentorMate, we have released this feature in [2.7.0](https://github.com/auth0/Auth0.Android/releases/tag/2.7.0). Do take a look into it. I am closing this issue for now please feel free to reopen it if you have any doubts.
